src/libraries/System.Private.CoreLib/src/System/Numerics/TotalOrderIeee754Comparer.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Numerics
+{
+    /// <summary>
+    /// Represents a comparison operation that compares floating-point numbers
+    /// with IEEE754 totalOrder semantic.
+    /// </summary>
+    /// <typeparam name="T">The type of the numbers to be compared, must be an IEEE754 floating-point type.</typeparam>
+    public class TotalOrderIeee754Comparer<T> : IComparer<T>, IEqualityComparer<T>
+        where T : IFloatingPointIeee754<T>?
+    {
+        /// <summary>
+        /// Compares two numbers with IEEE754 totalOrder semantic and returns
+        /// a value indicating whether one is less than, equal to, or greater than the other.
+        /// </summary>
+        /// <param name="x">The first number to compare.</param>
+        /// <param name="y">The second number to compare.</param>
+        /// <returns>
+        /// A signed integer that indicates the relative
+        /// values of <paramref name="x"/> and <paramref name="y"/>.
+        /// </returns>
+        /// <remarks>
+        /// IEEE754 specification defines totalOrder as a &lt;= semantic.
+        /// totalOrder(x,y) is true corresponds to the result of this method &lt;= 0.
+        /// </remarks>
+        public int Compare(T? x, T? y)
+        {
+            // IComparer contract is null < value
+
+            if (x is null && y is null)
+            {
+                return 0;
+            }
+            else if (x is null)
+            {
+                return -1;
+            }
+            else if (y is null)
+            {
+                return 1;
+            }
+
+            // If < or > returns true, the result satisfies definition of totalOrder too
+
+            if (x < y)
+            {
+                return -1;
+            }
+            else if (x > y)
+            {
+                return 1;
+            }
+            else if (x == y)
+            {
+                if (T.IsZero(x) && T.IsZero(y))
+                {
+                    if (T.IsNegative(x) && T.IsPositive(y))
+                    {
+                        return -1;
+                    }
+                    else if (T.IsPositive(x) && T.IsNegative(y))
+                    {
+                        return 1;
+                    }
+                    else
+                    {
+                        return 0;
+                    }
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+            else
+            {
+                // One or two of the values are NaN
+                // totalOrder defines that -qNaN < -sNaN < x < +sNaN < + qNaN
```

Comment:
What is the exact meaning of "we don't really have a concept of sNaN"? No corresponding representation in float/double/half?

Reply:
`sNaN` is a `NaN` representation where its typically expected to raise an `IEEE 754 floating-point exception` in most places its encountered.

.NET disables all `IEEE 754 floating-point exceptions` and so for us, `sNaN` is largely treated the same as `qNaN` and it will not be something a developer "typically" encounters.

We should still do the "correct" thing here with regards to sorting when such a value is encountered, however.